src/query/query_eval.rs

```diff
@@ -33,34 +33,74 @@ fn eval_latest<'a>(
 fn eval_lookup<'a>(
     index: &'a Index,
     lookup_field: LookupLhs,
-    value: &str,
+    value: LookupRhs,
 ) -> Result<Vec<&'a Packet>, QueryError> {
     Ok(index
         .packets
         .iter()
         .filter(|packet| match lookup_field {
-            LookupLhs::Id => packet.id == value,
-            LookupLhs::Name => packet.name == value,
-            LookupLhs::Parameter(param_name) =>
-                packet.parameter_equals(param_name, ParameterValue::String(value))
+            LookupLhs::Id => match value {
+                LookupRhs::String(str) => packet.id == str,
+                _ => false
+            },
+            LookupLhs::Name => match value {
+                LookupRhs::String(str) => packet.name == str,
+                _ => false
+            },
+            LookupLhs::Parameter(param_name) => packet.parameter_equals(param_name, &value)
         })
         .collect())
 }
 
+impl Packet {
+    #[allow(dead_code)]
+    fn parameter_equals(&self, param_name: &str, value: &LookupRhs) -> bool {
```

Comment:
this really begs for a comment or two explaining the idea I'm afraid. It might also be easier to read if the code arm of the match branches went into functions with descriptive names, and/or some helpers on the serde bit as it's really hard to scan atm

Reply:
Tried tidying that up a little by breaking the inner part (after we retrieve the parameter value) into another function. Sadly we're a little limited and can't break out each of the sections as we can't write a function signature which takes an enum variant! 

So I can't have e.g.

```
fn json_number_eq(json_val: serde_json::value::Value::Number, test_value: LookupRhs::Number) {
    if json_val.is_f64() {
        let test_number = serde_json::Number::from_f64(*test_val);
        match test_number {
            Some(number) => *json_val == number,
            None => false,
        }
    } else {
        *json_val == serde_json::Number::from(*test_val as i32)
    }
}
```